libdrgn/dwarf_index.c

```diff
@@ -206,16 +206,37 @@ static void
 drgn_dwarf_index_namespace_init(struct drgn_dwarf_index_namespace *ns,
 				struct drgn_dwarf_index *dindex)
 {
-	array_for_each(shard, ns->shards) {
-		omp_init_lock(&shard->lock);
-		drgn_dwarf_index_die_map_init(&shard->map);
-		drgn_dwarf_index_die_vector_init(&shard->dies);
-	}
+	ns->shards = NULL;
 	ns->dindex = dindex;
 	drgn_dwarf_index_pending_die_vector_init(&ns->pending_dies);
 	ns->saved_err = NULL;
 }
 
+static struct drgn_error *
+drgn_dwarf_index_namespace_shards_init(struct drgn_dwarf_index_namespace *ns)
+{
+	// Check once for init before acquiring to avoid lock contention once things are initialized.
+	if (ns->shards)
+		return NULL;
+	const size_t num_shards = 1 << DRGN_DWARF_INDEX_SHARD_BITS;
```

Comment:
Let's `#define DRGN_DWARF_INDEX_NUM_SHARDS (1 << DRGN_DWARF_INDEX_SHARD_BITS)` in `dwarf_index.h` and use that everywhere.